### PR TITLE
 libutil/logging: extend `internal-json` logger to make it more machine-readable

### PR DESCRIPTION
**Pull request by Ma27**
The new error-format is pretty nice from a UX point-of-view, however
it's fairly hard to parse the output e.g. for editor plugins such as
vim-ale[1] that use `nix-instantiate --parse` to determine syntax errors in
Nix expression files.

This patch extends the `internal-json` logger by adding the fields
`line`, `column` and `file` to easily locate an error in a file and the
field `raw_msg` which contains the error-message itself without
code-lines and additional helpers.

An exemplary output may look like this:

```
[nix-shell]$ ./inst/bin/nix-instantiate ~/test.nix --log-format minimal
{"action":"msg","column":1,"file":"/home/ma27/test.nix","level":0,"line":4,"raw_msg":"syntax error, unexpected IF, expecting $end","msg":"<full error-msg with code-lines etc>"}
```

[1] https://github.com/dense-analysis/ale

### CODE REVIEW

**Comment by Kloenk** [comment]
Is Jason a better name, because it's in a Jason format?

**Comment by Ma27** [comment]
Fair point. Can we agree on `minimal-json` though? That's IMHO better to make it distinguishable from `internal-json` (which I intentionally didn't choose as it basically puts the entire, pretty-printed error into a JSON).

cc @edolstra WDYT?

**Comment by edolstra** [comment]
I don't really understand why we need another JSON output format when we already have one. What information can't you get from the existing format?

**Comment by Ma27** [comment]
@edolstra the current json format returns a fully detailed message containing code-lines etc. This is IMHO fairly hard to parse for editor-plugins where I only need the line-number, line-column and message (which is provided in single JSON-fields in my new logger).

Also, we probably *want* to use the full-blown human-readable message in the `internal-json` logger (AFAIU) since that's how log-messages are received from a remote builder.

Finally, the `internal-json` logger is subject to change between releases (according to the man-pages) which would mean that plugins have to be modified for each new Nix release (in the worst case).

I see two options to take care of it:

* provide another logger for the purpose I described (which is currently the case in this PR)
* modify the `internal-json` logger to define (1) a somewhat stable structure and (2) expose error-msg (without code-lines), line-nr and line-column as additional json fields.

I'm fine with either of those options, so feel free to decide which you prefer.

**Comment by Ma27** [comment]
cc @bburdette who also might be interested in reviewing this as they implemented the beautiful new error-messages.

**Comment by edolstra** [comment]
I think it's better to improve the existing JSON logger than to add another one. The whole point of the JSON logger was that it's the last logger we'll ever need, since it should provide all information in a format that other tools can easily process.

> Finally, the internal-json logger is subject to change between releases

I don't think that should be the case. Having a log format that can change in arbitrary ways is not very useful. I think it's more that clients should be prepared to handle message types or fields being added or removed. (Conversely though, who says that the a new logger wouldn't be subject to change?)

> Also, we probably want to use the full-blown human-readable message in the internal-json logger (AFAIU) since that's how log-messages are received from a remote builder.

No, the daemon should send structured error messages to the client. Currently errors from the daemon don't get formatted very well. For instance, the daemon doesn't know the terminal width of the client. And it's hard for the client to add context to an error received from the daemon.

> modify the internal-json logger to define (1) a somewhat stable structure and (2) expose error-msg (without code-lines), line-nr and line-column as additional json fields.

I think this is the best approach. The code lines can probably be put in a separate JSON field which tools like IDEs can ignore.

**Comment by Ma27** [comment]
> Conversely though, who says that the a new logger wouldn't be subject to change?

Well, first of all it was explicitly stated in the man-pages. Also, error-information is only provided in the `msg` field which can be an arbitrary string (which is in *this* case a problem as we'd currently have to parse e.g. line-numbers etc. from it).

> I think this is the best approach. The code lines can probably be put in a separate JSON field which tools like IDEs can ignore.

That's also perfectly fine for me. I'll update the PR later today, thanks for your feedback!

**Comment by Ma27** [comment]
@edolstra updated accordingly %)

**Comment by Ericson2314** [comment]
Maybe take a look at what i wrote in https://github.com/NixOS/nix/issues/3841 ? I think this separation of rendering and data would also be good for machine-readable logging; currently we bake a lot of data in the hint which is conceptually separate, and should be actually separate for machine-readable logging.

**Comment by bburdette** [comment]
I like it!  I could see adding the trace info into the json format if at some point there's a tool that needs it.

**Comment by Ma27** [comment]
@bburdette fair point, added traces as well.
@Ericson2314 AFAIU this is something where we'd have to refactor the error-API first, right? After reading through the issue I'd generally agree with you, but this seems slightly out of scope for this PR (unless I misunderstood your point).
@edolstra anything else tbd? :)

**Comment by Ericson2314** [comment]
Yes I suppose this changes are independent of this PR.

File: src/libutil/logging.cc [hunk 1]
```diff
@@ -184,6 +184,33 @@ struct JSONLogger : Logger {
...
+
+        if (ei.errPos.has_value() && (*ei.errPos)) {
+            json["line"] = ei.errPos->line;
+            json["column"] = ei.errPos->column;
+            json["file"] = ei.errPos->file;
+        } else {
+            json["line"] = -1;
+            json["column"] = -1;
+            json["file"] = nullptr;
+        }
```
Comment (Ericson2314) [inline src/libutil/logging.cc:197]: I don't think `-` is a great sentinel value. I rather do one of:

 1. Make suboject, stored at `pos`, which is nullable.
 2. Use `null` for each as the sentinel value
 3. Skip fields altogether rather than using sentinel value

in decreasing order of preference.

Comment (Ma27) [reply src/libutil/logging.cc]: @Ericson2314 I decided to set `null` by default here since AFAIU the `pos` struct *can* be undeclared, so with this we have a fair sentinel value and are on the safe side IMHO :)

**Comment by Ma27** [comment]
afaics all comments are resolved. Anything else to add @edolstra?

**Comment by Ma27** [comment]
ping @edolstra anything else tbd to get this merged?

**Comment by Ma27** [comment]
ping @edolstra @Ericson2314
